### PR TITLE
tests: add solidity test suites

### PR DESCRIPTION
**Pull request by sohkai**
## Description

Adds a set of solidity test suites against ethermint, mostly pulled from existing repos. Note that these are increasingly difficult, and currently, the following do not pass:

- `initializable-buidler`: likely due to how buidler uses `eth_estimateGas`, as `initializable` (using truffle) runs fine. Causes an RPC panic, after which all RPC requests fail (also pretty awful behaviour for a production-grade RPC API).
- `proxy`: appears to fail when trying to use `sendTransaction()` with `value`
- `staking`: largest of them all (but the simplest of our production-grade contract suites). Immediately panics and fails :shrug:.

Note that we would like to have [`aragonOS`](https://github.com/aragon/aragonOS), [`aragon-court`](https://github.com/aragon/aragon-court), and all of the [`aragon-apps`](https://github.com/aragon/aragon-apps)' tests passing.

Having `staking` pass against `ethermint` would be a great first step towards those larger test suites (2-10x larger).

Also note that I've scrubbed any revert strings for now, as ethermint apparently does not follow geth 1.9.5+'s behaviour of returning the revert string on `eth_call`s.
______

For admin use:

- [ ] Added appropriate labels to PR (ex. `WIP`, `R4R`, `docs`, etc)
- [ ] Reviewers assigned
- [ ] Squashed all commits, uses message "Merge pull request #XYZ: [title]" ([coding standards](https://github.com/tendermint/coding/blob/master/README.md#merging-a-pr))

### CODE REVIEW

File: tests-solidity/init-test-node.sh [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+echo $USER1_MNEMONIC | ethermintcli keys add $USER1_KEY --recover
+echo $USER2_MNEMONIC | ethermintcli keys add $USER2_KEY --recover
+
+# Set moniker and chain-id for Ethermint (Moniker can be anything, chain-id must be an integer)
+ethermintd init $MONIKER --chain-id $CHAINID
+
+# Allocate genesis accounts (cosmos formatted addresses)
+ethermintd add-genesis-account $(ethermintcli keys show $VAL_KEY -a) 1000000000000000000000photon,10000000000000000stake
+ethermintd add-genesis-account $(ethermintcli keys show $USER1_KEY -a) 1000000000000000000000photon,10000000000000000stake
+ethermintd add-genesis-account $(ethermintcli keys show $USER2_KEY -a) 1000000000000000000000photon,10000000000000000stake
```
Comment (fedekunze) [inline tests-solidity/init-test-node.sh]:
```suggestion
ethermintd add-genesis-account $(ethermintcli keys show $VAL_KEY -a) 1000000000000000000000aphoton,10000000000000000stake
ethermintd add-genesis-account $(ethermintcli keys show $USER1_KEY -a) 1000000000000000000000aphoton,10000000000000000stake
ethermintd add-genesis-account $(ethermintcli keys show $USER2_KEY -a) 1000000000000000000000aphoton,10000000000000000stake
```

File: tests-solidity/init-test-node.sh [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+
+# Collect genesis tx
+ethermintd collect-gentxs
+
+# Enable faucet
+cat  $HOME/.ethermintd/config/genesis.json | jq '.app_state["faucet"]["enable_faucet"]=true' >  $HOME/.ethermintd/config/tmp_genesis.json && mv $HOME/.ethermintd/config/tmp_genesis.json $HOME/.ethermintd/config/genesis.json
+
+echo -e '\n\ntestnet faucet enabled'
+echo -e 'to transfer tokens to your account address use:'
+echo -e "ethermintcli tx faucet request 100photon --from $VAL_KEY\n"
```
Comment (fedekunze) [inline tests-solidity/init-test-node.sh]:
```suggestion
echo -e "ethermintcli tx faucet request 100aphoton --from $VAL_KEY\n"
```

**Review by fedekunze — APPROVED** [review]
Thanks for the tests! Ideally, we should add them as Github action workflows but since this PR is already large I'll merge it as is and then work on a follow-up

**Comment by sohkai** [comment]
Thanks! I was also asking @noot if we had some sort of "instant seal" mode to mine a block upon receiving a transaction to make these tests go faster—waiting 7s for each transaction makes these run quite slowly 😄

**Pull request merged**